### PR TITLE
#351 new progress bar component

### LINKED ISSUE #351: Create component ProgressBarNEW to use new Full structure  and use revalidation logic

### PR DESCRIPTION
**Pull request by nmadruga**
Fixes #351 

~~Branched of #366 please merge it first before this one!~~ DONE

### Changes
- Use new `FullStructure` to display on the `ProgressBar` new component, now each Page has a progress and no longer needs to deduce the page progress on the go.
- Simplified the component (compared to old `ProgressBar`) to be easier to read, there is an alternative to using `Grid` commented, but that needs a bit tweaking to get rid of the background on `Label`...
- Change pages on **non linear** applications, ~~still has a TODO to deal with **linear** which is pending on checking with @andreievg what was the approach decided (it has been removed from Application Flow chart).~~
- [In progress] Adding validation to change to another page on **linear** applications

### CODE REVIEW

**Comment by CarlosNZ** [comment]
Cool, can you pull in the latest `master` (and maybe the recent changes to #366) which has the PageElements in it, so we can test with actual content on the page?

**Comment by nmadruga** [comment]
> Cool, can you pull in the latest master (and maybe the recent changes to #366) which has the PageElements in it, so we can test with actual content on the page?

Sure! Sorry, I thought I had done it on 22574a7

**Comment by CarlosNZ** [comment]
> Sure! Sorry, I thought I had done it on [22574a7](https://github.com/openmsupply/application-manager-web-app/commit/22574a7a2cece49deb594ec54ee06b21845ef23c)

Yeah, it doesn't have the PageElements in it yet, need to pull that from latest master.

**Comment by nmadruga** [comment]
Oh! Oops, half way done. I'll also re-merge #366 in this one with latest change to also be used to on tests. 1 minute.

**Comment by nmadruga** [comment]
Done 👍

File: src/components/Application/ProgressBarNEW.tsx
Comment (CarlosNZ) [inline src/components/Application/ProgressBarNEW.tsx:42]: I notice that the styling/layout of the new one is a bit different to the old one. Is that intentional?

**Review by CarlosNZ — COMMENTED** [review]
Cool, it's working well. I notice that the Page links are not yet clickable -- are you waiting for the revalidation logic to be finalised?

Also --have only had a really quick look at the code. I think you need to merge `master` into `#362` branch, because this PR is showing up with all the other changes in master (like the list, pageELements etc) as diffs since they're not in #362 yet, which this is branched off.

**Comment by nmadruga** [comment]
> Cool, it's working well. I notice that the Page links are not yet clickable -- are you waiting for the revalidation logic to be finalised?

They are clickable only for **non linear** applications

**Comment by CarlosNZ** [comment]
> They are clickable only for **non linear** applications

Shouln't they also be clickable for linear applications if it's a page I'm allowed to go to? If I'm up to page 3, and I go back and have a look at Page 1, I should be able to click the "Page3" link to get back to where I was, even with Linear.

**Comment by CarlosNZ** [comment]
> Shouln't they also be clickable for linear applications if it's a page I'm allowed to go to? If I'm up to page 3, and I go back and have a look at Page 1, I should be able to click the "Page3" link to get back to where I was, even with Linear.

Just checked and I can do this as expected in the current version.

**Comment by nmadruga** [comment]
> Shouln't they also be clickable for linear applications if it's a page I'm allowed to go to? If I'm up to page 3, and I go back and have a look at Page 1, I should be able to click the "Page3" link to get back to where I was, even with Linear.

Yes. That was a very barebones solution really I haven't add much of the logic/layout to make it work for the linear application yet, since I was waiting on the existing PRs to get merged first. Then it would be easier for me to implement what is missing for linear applications. 

But that's a good point. What do you think is best: allow any Page to be clickable - just not allow changing to the page if the validation fails or only enabling the pages that have been visited (as previous logic).

File: src/components/Application/ProgressBarNEW.tsx
Comment (andreievg) [inline src/components/Application/ProgressBarNEW.tsx:63]: I think we would use firstStrictInvalidPageFor this (if linear)

File: src/containers/Application/ApplicationPage.tsx
Comment (andreievg) [inline src/containers/Application/ApplicationPage.tsx:52]: Would prefer for progress component to do within itself

File: src/components/Application/ProgressBarNEW.tsx
Comment (andreievg) [inline src/components/Application/ProgressBarNEW.tsx:46]: Hmm, don't we want to show error if it's not completed but something is defiantly invalid ?

**Review by andreievg — CHANGES_REQUESTED** [review]
A few minor comments.

Styles are different to the original one (but can be done in new issue)

File: src/components/Application/ProgressBarNEW.tsx
Comment (andreievg) [reply src/components/Application/ProgressBarNEW.tsx:46]: Ohh looks like it does wen page refresh, never mind

File: src/components/Application/ProgressBarNEW.tsx
Comment (nmadruga) [reply src/components/Application/ProgressBarNEW.tsx:42]: Is this for page's icons? I changes the size which needs to be corrected. Or you noticed some other changes?

File: src/components/Application/ProgressBarNEW.tsx
Comment (nmadruga) [reply src/components/Application/ProgressBarNEW.tsx:63]: Have discussed briefly with Carl, maybe we shouldn't disable sections/pages just because since they are still links when hoovering it display the icon as clickable, is just the label that is changing. And when the user clicks on the a page ahead (on linear applications) it should just run the validation and display errors on the current page. 

So changing now to have all pages/sections enabled. Seems good?

Comment (andreievg) [reply src/components/Application/ProgressBarNEW.tsx:63]: yep

**Review by andreievg — APPROVED** [review]

File: src/containers/Application/ApplicationPage.tsx
Comment (nmadruga) [reply src/containers/Application/ApplicationPage.tsx]: Done in big commit: 32bf4f0

**Comment by nmadruga** [comment]
@andreievg please have another look on the changes done here to also use the validation method for linear applications. I think that's all what was missing now. Let me know of any suggestions to it (has also changes in issue #374)

**Review by andreievg — COMMENTED** [review]
@nmadruga I've noticed a few issues (not just with progress bar, but other things, and it was much quicker to make code changes and a PR (https://github.com/openmsupply/application-manager-web-app/pull/384). Happy for you to merge this one if the other one is approved rebased to master.

I wanted to make a PR instead of suggested changes branch, so you and Carl can comment on it

**Comment by nmadruga** [comment]
Hey @andreievg I've done a quick pass-through your PR on Saturday but will do a proper review today. So before merging this one we would be waiting on the next one to be approved and ready to merge. Correct?

**Pull request merged**
